File: Trabalho 1/A1_5.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <limits>
#include "A1_1.h"

using namespace std;

template <typename T>
vector<vector<double>> FloydWarshall(const Grafo<T>& G) {
    int n = G.getQuantidadeVertices();
    const double INF = numeric_limits<double>::infinity();
    
    vector<vector<double>> D(n, vector<double>(n, INF));

    // Inicialização
    for (int u = 0; u < n; ++u) {
        D[u][u] = 0;
        Aresta* atual = G.getListaAdjacencia(u);
        while (atual != nullptr) {
            D[u][atual->destino] = atual->peso;
            atual = atual->proxima;
        }
    }

    // Algoritmo de Floyd-Warshall
    for (int k = 0; k < n; ++k) {
        for (int u = 0; u < n; ++u) {
            for (int v = 0; v < n; ++v) {
                if (D[u][k] != INF && D[k][v] != INF) {
                    D[u][v] = min(D[u][v], D[u][k] + D[k][v]);
                }
            }
        }
    }

    return D;
}

void imprimirDistancias(const vector<vector<double>>& D) {
    int n = D.size();
    for (int i = 0; i < n; i++) {
        cout << (i) << ":";
        for (int j = 0; j < n; j++) {
            if (D[i][j] == numeric_limits<double>::infinity()) {
                cout << "INF";
            } else {
                cout << D[i][j];
            }
            if (j < n-1) cout << ",";
        }
        cout << endl;
    }
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        cerr << "Uso: " << argv[0] << " <arquivo_entrada>" << endl;
        return -1;
    }

    ifstream entrada(argv[1]);
    if (!entrada.is_open()) {
        cerr << "Erro ao abrir o arquivo: " << argv[1] << endl;
        return 1;
    }

    Grafo<string> g;
    g.ler(entrada);
    
    auto D = FloydWarshall(g);
    imprimirDistancias(D);

    return 0;
}

File: Trabalho 1/A1_5.py
import sys
from typing import List, Tuple, Optional, Dict
from collections import defaultdict
from A1_1 import Grafo, Vertice, Aresta

INF = float('inf')

def floyd_warshall(grafo):
    n = grafo.qtdVertices()

    D = [[INF] * n for _ in range(n)]

    for u in grafo.getArestas():
            D[u.origem.indice-1][u.destino.indice-1] = int(u.peso)
            D[u.destino.indice-1][u.origem.indice-1] = int(u.peso)

    # Algoritmo de Floyd-Warshall
    for k in range(n):
        for u in range(n):
            for v in range(n):
                if D[u][v] != INF:
                    if D[u][k] != INF and D[k][v] != INF:
                        D[u][v] = min(D[u][v], D[u][k] + D[k][v])
                else:
                    if D[u][k] != INF and D[k][v] != INF:
                        D[u][v] = D[u][k] + D[k][v]

    return D


if __name__ == "__main__":
    if len(sys.argv) != 2:
        
        print(f"Uso: {sys.argv[0]} <arquivo_grafo>")
        sys.exit(1)

    G = Grafo()
    G.ler(sys.argv[1])

    D = floyd_warshall(G)
    
    n = G.qtdVertices()
    for u in range(n):
        print(f"{u+1}:{','.join(map(str, D[u]))}\n")